File: lecture_notes/10 refactoring/improvements.py
import json

class Shift:
  def __init__(self,offset):
    self.offset = offset

  def move(self,point):
    return {
    'x': point['x'] + self.offset['dx'], 
    'y': point['y'] + self.offset['dy'],
    }

parsers = {
  type(''): json.loads, 
  type({}): lambda dict:dict
}

def conditional_parse(item):
  parser = parsers[type(item)]
  return parser(item)

polyline = [
  {'x': 0, 'y': 0},
  {'x' : 10, 'y' : 10},
  '{"x": 20, "y": 20}',
  {'x': 30, 'y': 30},
]

shift = Shift({ 'dx': 10, 'dy': -5 })
parsed = map(conditional_parse, polyline)
path = map(lambda point:shift.move(point), parsed)
print(path)

File: lecture_notes/10 refactoring/improvements.js
'use strict';

const move = (offset) => (point) => {
  point.x += offset.x;
  point.y += offset.y;
  return point;
};

const conditionalParse = (item) => {
  if (typeof item === 'object') return item;
  return JSON.parse(item);
};

const polyline = [
  { x: 0, y: 0 },
  { x: 10, y: 10 },
  '{ "x": 20, "y": 20 }',
  { x: 30, y: 30 },
];

const offset = move({ x: 10, y: -5 });
const parsed = polyline.map(conditionalParse);
const path = parsed.map(offset);
console.log({ path });
